### PR TITLE
Cast only floating types with IPUs

### PR DESCRIPTION
**Pull request by rohitgr7**
## What does this PR do?

Reported by graphcore team.

## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

cc @borda @carmocca @justusschock @awaelchli @akihironitta @rohitgr7

### CODE REVIEW

**Comment by rohitgr7** [comment]
pulling it out of the draft to run IPU tests.

**Review by carmocca — APPROVED** [review]

File: src/pytorch_lightning/strategies/ipu.py [hunk 1]
```diff
@@ -272,8 +272,13 @@ def to_tensor(x):
     def batch_to_device(self, batch: Any, device: Optional[torch.device] = None, dataloader_idx: int = 0) -> Any:
         # This override is necessary because the cast must occur before the data
         # is moved to the device to prevent wasteful host->device copies.
+        def fp_to_half(tensor):
```
Comment (carmocca) [inline src/pytorch_lightning/strategies/ipu.py:275]: Should add a CHANGELOG entry

Comment (rohitgr7) [reply src/pytorch_lightning/strategies/ipu.py:275]: yes.. I was waiting for another Changelog PR to merge..

File: src/pytorch_lightning/strategies/ipu.py [hunk 2]
```diff
@@ -272,8 +274,13 @@ def to_tensor(x):
...
         # This override is necessary because the cast must occur before the data
         # is moved to the device to prevent wasteful host->device copies.
+        def fp_to_half(tensor: Tensor) -> Tensor:
+            if torch.is_floating_point(tensor):
+                return tensor.half()
+            return tensor
+
         if self.precision_plugin.precision in (PrecisionType.MIXED, PrecisionType.HALF):
-            batch = apply_to_collection(batch, Tensor, function=Tensor.half)
+            batch = apply_to_collection(batch, Tensor, function=fp_to_half)
```
Comment (awaelchli) [inline src/pytorch_lightning/strategies/ipu.py:283]:
```suggestion
            batch = apply_to_collection(batch, Tensor, function=self.batch_to)
```

**Review by awaelchli — APPROVED** [review]

File: src/pytorch_lightning/strategies/ipu.py (hunk 2 above)
Comment (rohitgr7) [reply src/pytorch_lightning/strategies/ipu.py:283]: that's a different deprecated class that is not used anymore

Comment (awaelchli) [reply src/pytorch_lightning/strategies/ipu.py:283]: true! sorry, didn't unfold the wrapped code.

**Review by kaushikb11 — APPROVED** [review]

**Review by justusschock — APPROVED** [review]

**Pull request merged**